swim/heal_partition.go

```diff
@@ -0,0 +1,127 @@
+// Copyright (c) 2016 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package swim
+
+import "time"
+
+// AttemptHeal attempts to heal a partition between the node and the target.
+// It is the responsibillity of the caller to detect that the node is indeed
+// separated from the target.
+//
+// The algorithm to heal the partition works as follows:
+// The goal is to first let every node reincarnate, we do this by declaring nodes
+// suspect. For members that are faulty in partition B, we fabricate updates with
+// a suspect in stead of faulty status and disseminate them through A. We also do
+// the reverse. We then create a connection from partition A to B and visa versa.
+// We do this by reincarnating this node and sending the respective update to the
+// target. The target has also been reincarnated due to previous actions, this
+// update is now applied on this node's membership to create the connection from
+// partition B to partition A.
+//
+// The heal alogrithm in more detail:
+// 1. send join request to the target, this tells us whether a partition exists
```

Comment:
Not immediately clear how this tells us the partition exists, and it would be good to explain. "... a partition exists: we can see that members which are alive in the current node are declared faulty in target" ? Is that somewhat accurate?